AsyncDisplayKit/Details/ASCollectionViewFlowLayoutInspector.m

```diff
@@ -0,0 +1,121 @@
+/* Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import <UIKit/UIKit.h>
+
+#import "ASCollectionViewFlowLayoutInspector.h"
+
+#import "ASCollectionView.h"
+
+@implementation ASCollectionViewFlowLayoutInspector {
+  BOOL _delegateImplementsReferenceSizeForHeader;
+  BOOL _delegateImplementsReferenceSizeForFooter;
+}
+
+#pragma mark - Accessors
+
+- (instancetype)initWithCollectionView:(ASCollectionView *)collectionView flowLayout:(UICollectionViewFlowLayout *)flowLayout;
+{
+  self = [super init];
+
+  if (flowLayout == nil) {
+    return nil;
+  }
+
+  if (self != nil) {
+    [self cacheSelectorsForCollectionView:collectionView];
+    _layout = flowLayout;
+  }
+  return self;
+}
+
+- (void)cacheSelectorsForCollectionView:(ASCollectionView *)collectionView
+{
+  if (collectionView == nil) {
+    _delegateImplementsReferenceSizeForHeader = nil;
+    _delegateImplementsReferenceSizeForFooter = nil;
+  } else {
+    _delegateImplementsReferenceSizeForHeader = [[self delegateForCollectionView:collectionView] respondsToSelector:@selector(collectionView:layout:referenceSizeForHeaderInSection:)];
+    _delegateImplementsReferenceSizeForFooter = [[self delegateForCollectionView:collectionView] respondsToSelector:@selector(collectionView:layout:referenceSizeForFooterInSection:)];
+  }
+}
+
+#pragma mark - ASCollectionViewLayoutInspecting
+
+- (ASSizeRange)collectionView:(ASCollectionView *)collectionView constrainedSizeForSupplementaryNodeOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath
+{
+  CGSize constrainedSize;
+  CGSize supplementarySize = [self sizeForSupplementaryViewOfKind:kind inSection:indexPath.section collectionView:collectionView];
+  if (_layout.scrollDirection == UICollectionViewScrollDirectionVertical) {
+    constrainedSize = CGSizeMake(collectionView.bounds.size.width, supplementarySize.height);
+  } else {
+    constrainedSize = CGSizeMake(supplementarySize.height, collectionView.bounds.size.height);
+  }
+  return ASSizeRangeMake(CGSizeZero, constrainedSize);
```

Comment:
@levi - This is important...I think it may not be quite correct.  I believe this had to be changed for ASTableView recently.

@nguyenhuy, @rcancro — was the discovery / conclusion that we needed to provide a minimum size of the table width + 0.0 height in order for certain layout spec calculations to work correctly?  If so, we should probably do something similar here where depending on the scroll direction, the appropriate dimension of the minimum size is pegged.

Also important: I'm pretty sure that scrollDirection for collectionView can be BOTH vertical and horizontal at the same time.  Presumably in this case the minimum size of CGSizeZero is correct.